community/rfcs/24-07-11-OPEA-Agent.md

```diff
@@ -0,0 +1,278 @@
+
+ Advantage and limitation:   
+ * User can create a session thread memorizing previous conversation as long-term memory. And Human-In-Loop agent will only works use this API.
+ * User client application may need codes change to work with this new API.
+ * openAI assistant API is tagged with ‘beta’, not stable
+
+ ```
+ # assistants API is used to create agent runtime instance with a set of tool / append addition instructions
+ - "/v1/assistants": {
+                         "instructions": str,
+                         "name": str,
+                         "tools": list
+                     }
+ 
+ # threads API is to used maintain conversation session with one user. It can be resumed from previous, can tracking long term memories.
+ - "/v1/threads/ ": { # empty is allowed }
+ 
+ 
+ # threads messages API is to add a task content to thread_1 (the thread created by threads API)
+ - "/v1/threads/thread_1/messages": {
+                             "role": str,
+                             "content": str
+                         }
+ 
+ # threads run API is to start to execute agent thread using run api
+ 
+ - "/v1/threads/thread_1/runs": {
+                             'assistant_id': str,
+                             'instructions': str,
+                         }
+ ```
+
+ ### Part 2. 'Agent' genAI Component definition
+
+ 'Agent' genAI Component is regarded as the resource management unit in “Agent” design.  It will be launched as one microservice and can be instantiated as ‘Agent’, ‘Planner’ or ‘Executor’ according to configuration. Tools will be registered to 'Agent' microservice during launch or runetime.
+
+ ![image](https://github.com/user-attachments/assets/38e83fa4-57d8-4146-9061-e5153472b5f4)
+
+ #### SPEC for any agent Role - agent, planner, executor
+ ```
+ "/v1/chat/completions": {
+                         "model": str,
+                         "messages": list,
+                         "tools": list,
+                     }
+ "/v1/assistants": {
+                         "instructions": str,
+                         "name": str,
+                         "tools": list
+                     }
+ "/v1/threads/: {}
+ "/v1/threads/thread_1/runs": {
+                             'assistant_id': str,
+                             'instructions': str,
+                         }
+ "/v1/threads/thread_1/messages": {
+                             "role": str,
+                             "content": str
+                         }
+ ```
+
+  #### Agent Role microservice definition - 'Agent': 
+  A complete implementation of Agent, which contains LLM endpoint as planner, strategy algorithm for plan execution, Tools, and database handler to keep track of historical state and conversation. 
+
+  configuration:
+  ```
+  strategy: choices([react, planexec, humanInLoopPlanExec])
+  require_human_feedback: bool
+  llm_endpoint_url: str
+  llm_engine: choices([tgi, vllm, openai])
+  llm_model_id: str
+  recursion_limit: int
```

Comment:
any description for these field?

Reply:
Good suggestion, recursion_limit is not a must_to_provide argument, default will be 5. This keyword is used to ask agent to break after several tests. For example:

If we asked agent to get today's intel stock price but not providing right tools, agent will not get the exact answer and fall into a dead loop, in which the Agent will try to fix its prompt and plan back and forth). By adding 'recursion_limit', if one agent can't get answer after 'recursion_limit' times of tests, agent will direct return to user with no answer.

Reply:
Is this resolved?